src/include/common/worker_pool.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+#include <atomic>
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <iostream>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <string>
+#include <thread>  // NOLINT
+#include <utility>
+#include <vector>
+
+namespace terrier::common {
+
+using TaskQueue = std::queue<std::function<void()>>;
+
+/**
+ * A worker pool that maintains a group of worker threads and a task queue.
+ *
+ * As soon as there is a task in the task queue, a worker thread will be
+ * assigned to run that task. A task must be a function that takes no argument.
+ * After a worker finishes a task, it will eagerly try to get a new task.
+ *
+ * This pool is restartable, meaning it can be started again after it has been
+ * shutdown. Calls to Startup() and Shutdown() are thread-safe.
+ */
+class WorkerPool {
+ public:
+  /**
+   * Initialize the worker pool. Once the number of worker is set by the constructor
+   * it cannot be changed.
+   *
+   * After initialization of the worker pool with a given task queue, worker threads
+   * do NOT start running. Need call StartUp() to start working on tasks.
+   *
+   * @param pool_name the name of the worker pool
+   * @param num_workers the number of workers in this pool
+   * @param task_queue a queue of tasks
+   */
+  WorkerPool(std::string pool_name, uint32_t num_workers, TaskQueue task_queue)
```

Comment:
I don't think this has to be fixed size? Granted it provides some immutability, but:

- we want to be able to tune these things at runtime.
- it doesn't seem to be very hard to change the number of threads in the thread pool (in either direction)
- (not really a good reason, but the test thread pool kind of need a flexible number of threads)